### PR TITLE
feat(azure): add `azure` plugin

### LINKED ISSUE #8847: Support for Azure cli
**Is your feature request related to a problem? Please describe.**
ohmyzsh is missing support for the tool azure-cli
https://docs.microsoft.com/en-us/cli/azure/

**Describe the solution you'd like**
A plugin like the one in https://github.com/ohmyzsh/ohmyzsh/tree/master/plugins/aws with similar utilities (profiles --> subscriptions)

**Describe alternatives you've considered**
It looks like people have been living without a plugin for azure for a while, but is that fun ?

### PR DESCRIPTION
**Pull request by jrottenberg**
## Standards checklist:

- [x] The PR title is descriptive.
- [x] The PR doesn't replicate another PR which is already open.
- [x] I have read the contribution guide and followed all the instructions.
- [x] The code follows the code style guide detailed in the wiki.
- [x] The code is mine or it's from somewhere with an MIT-compatible license.
- [x] The code is efficient, to the best of my ability, and does not waste computer resources.
- [x] The code is stable and I have tested it myself, to the best of my abilities.

## Changes:

- add an Azure plugin

### New aliases

* `az_subscriptions`: lists the available subscriptions.
* `azgs`: gets the current value of `$azure_subscription`.
* `azss [<subscription>]`: sets the `$azure_subscription`.

## Other comments:

Fixes  #8847

### CODE REVIEW

**Comment by jrottenberg** [comment]
I've worked with my buddy @hagridaaron on this one
It's our first PR against this project, we read https://github.com/ohmyzsh/ohmyzsh/blob/master/CONTRIBUTING.md but let us know if we missed anything

**Comment by awasum** [comment]
Please squash the commit into one so the git history is clean

**Comment by awasum** [comment]
@jrottenberg , if you have a co-author, you can squash and add the co-author like so: https://help.github.com/en/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors

**Comment by jrottenberg** [comment]
absolutely.

**Review by 53jk1 — APPROVED** [review]

**Comment by CoaxVex** [comment]
Can this be merged? I'd love this feature. Thx!

**Comment by oooookk7** [comment]
Please merge this. Need this. Thank you!

**Comment by mcornella** [comment]
@jrottenberg does `az` provide completion support for `az account set --subscription`? If so, a simple alias would be better.

**Comment by jrottenberg** [comment]
Hello !

There is completion support but it relies heavily on bash : https://github.com/Azure/azure-cli/blob/dev/az.completion

Not sure how to extract / alias it for `az  account set --subscription`

**Comment by mcornella** [comment]
Zsh has `bashcompinit` to read bash completion files, which is automatically called by OMZ already. Regardless, you'd do
```zsh
autoload -U +X bashcompinit && bashcompinit
source /path/to/az.completion
```

That last line looks like it's dependent on how you installed `az`. For example, using `apt`, the file is put in `/etc/bash_completion.d/azure-cli`. Perhaps it would be better to load it from the plugin.

**Comment by jrottenberg** [comment]
Thank you for the feedback, yeah that makes sense :

1.  I'd kill the whole function, since it can be aliased, the only difference is the list now contains the subscriptionId, along with their name, but I'd call that a feature ;-)
1. Not sure to follow you, I'm looking for the most probable location for the plugin, with a fallback to `/etc/bash_completion.d/azure-cli`. You meant to "not try to load the completion altogether" ?

**Comment by doubleyewdee** [comment]
Does this still need testers? I'd be happy to validate some functionality :)

File: plugins/azure/azure.plugin.zsh
Comment (tmoschou) [inline plugins/azure/azure.plugin.zsh:21]: Convention seems to be `<plugin>_prompt_info()`
```suggestion
function azure_prompt_info() {
```

And would probably need to add the stub at `lib/prompt_info_functions.zsh`

File: plugins/azure/README.md
Comment (tmoschou) [inline plugins/azure/README.md:37]:
```suggestion
- ZSH_THEME_AZURE_PREFIX: sets the prefix of the azure_subscription. Defaults to `<az:`.
```

File: plugins/azure/README.md
Comment (tmoschou) [inline plugins/azure/README.md:39]:
```suggestion
- ZSH_THEME_AZURE_SUFFIX: sets the suffix of the azure_subscription. Defaults to `>`.
```

File: plugins/azure/README.md
Comment (tmoschou) [inline plugins/azure/README.md:30]:
```suggestion
NOTE : because azure keeps the state of ative subscription in ${AZURE_CONFIG_DIR:-$HOME/.azure/azureProfile.json}, the prompt command requires `jq` to be enabled. If jq is not in the path the prompt will show nothing
```

File: plugins/azure/azure.plugin.zsh
Comment (tmoschou) [inline plugins/azure/azure.plugin.zsh:23]:
```suggestion
  # azgs is too expensive, if we have jq, we enable the prompt
```

File: plugins/azure/azure.plugin.zsh
Comment (tmoschou) [inline plugins/azure/azure.plugin.zsh:24]: I think this is preferred?

```suggestion
  (( $+commands[jq] )) || return 1
```

File: plugins/azure/azure.plugin.zsh
Comment (tmoschou) [inline plugins/azure/azure.plugin.zsh:25]: This should avoid an additional subprocess
```suggestion
  azgs=$(jq -r '.subscriptions[] | select(.isDefault==true) .name' ${AZURE_CONFIG_DIR:-$HOME/.azure/azureProfile.json})
```

File: plugins/azure/azure.plugin.zsh
Comment (tmoschou) [inline plugins/azure/azure.plugin.zsh:26]:
```suggestion
  echo "${ZSH_THEME_AZURE_PREFIX:=<az:}${azgs}${ZSH_THEME_AZURE_SUFFIX:=>}"
```

**Review by tmoschou — COMMENTED** [review]
Thanks for raising a PR. I'm not a contributor, but a user also interested in an azure plugin. Here are some comments

File: plugins/azure/azure.plugin.zsh
Comment (tmoschou) [inline plugins/azure/azure.plugin.zsh:30]: As a side, perhaps a question for the ohmyzsh maintainers, should new plugins generally be allowed to munge the prompt? It seems only three plugins actually do (`aws`, `fossil`, and `git-prompt` - though the last is specifically a plugin for the prompt). The rest rather document the `*_prompt_info()` and leave it up to the user and/or themes to customise. For instance [plugins/jenv/README.md](https://github.com/ohmyzsh/ohmyzsh/tree/0ab87c26c17171ae6162ff379a0c704fa57dff2e/plugins/jenv#theme-example), [plugins/virtualenv/README.md](https://github.com/ohmyzsh/ohmyzsh/tree/0ab87c26c17171ae6162ff379a0c704fa57dff2e/plugins/virtualenv#virtualenv)

Perhaps that munging behaviour is legacy in those plugins, and not removed for compatibility reasons?

**Comment by jrottenberg** [comment]
Thank you very much @tmoschou ! I didn't know you could push patch to the branch like that in Github, super nice. All look correct, but I want to test some locally before to commit.

File: plugins/azure/azure.plugin.zsh
Comment (jrottenberg) [reply plugins/azure/azure.plugin.zsh:22]: Very interesting, we need a file check to make sure the user has a valid session, at the same time the fallback is not too bad:

```
$ az account list  --all --output tsv --query '[*].name'
Please run "az login" to access your accounts.
```

I'll add a 2>/dev/null

File: plugins/azure/azure.plugin.zsh
Comment (jrottenberg) [reply plugins/azure/azure.plugin.zsh]: Yeah unless I'm missing something, I don't see a canonical way to update a prompt, I"m using POWERLEVEL9K (damn deprecated too, needs 10k)

Comment (jrottenberg) [reply plugins/azure/azure.plugin.zsh]: oh wow, yeah it seems to be a theme decision, the powerlevel10k integration is pretty nice (loads the current subscriptions in the right prompt only if I type `az`)

I'll keep `azgs` as variable to use regardless in a RPROMPT, but that simplifies a lot.

File: plugins/azure/azure.plugin.zsh
Comment (jrottenberg) [reply plugins/azure/azure.plugin.zsh:30]: Absolutely, the plugin won't make assumption anymore on where it will show up, it will just load aliases, prepare functions and delegate the prompt to PROMPT or the theme explicitly.

**Comment by sdavis3** [comment]
Can we please get this merged into the master branch?

**Comment by Etienne-Schmitt** [comment]
Is there any news about this?

**Comment by carlosala** [comment]
Thanks everyone for the contributions, and sorry for being so late!

**Pull request merged**
